### PR TITLE
build: 3 create service template

### PR DESCRIPTION
**Pull request by foobazdev**
1. Remove old primary template
2. Create primary Users microservice application
3. Implement database connection for Users service
4. Create common docker compose file for project
5. Create primary shared eslint config

### CODE REVIEW

File: backend/users/Dockerfile
Comment (M-Ziusko) [inline backend/users/Dockerfile:1]: Use docker init to imrove this file

File: backend/users/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+npx prisma migrate deploy
+```
+
+## API
+
+- Users `/api/users`
+  - GET `/api/users` returns all users
+  - GET `/api/users:id` returns user by id
+  - POST `/api/users` create user 
+  - PATCH `/api/users` update user
```
Comment (M-Ziusko) [inline backend/users/README.md:63]: Better to use swagger for this purpose

File: backend/users/src/prisma/prisma.service.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { Injectable, OnModuleInit } from '@nestjs/common';
+import { PrismaClient } from '@prisma/client';
+
+@Injectable()
+export class PrismaService extends PrismaClient implements OnModuleInit {
+  async onModuleInit() {
+    await this.$connect();
+  }
+}
```
Comment (M-Ziusko) [inline backend/users/src/prisma/prisma.service.ts:9]: Add onModuleDestroy with disconnect

File: backend/users/src/users/users.service.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import { CreateUserDto } from './dto/create-user.dto';
+import { UpdateUserDto } from './dto/update-user.dto';
+import { PrismaService } from 'src/prisma/prisma.service';
+import * as bcrypt from 'bcrypt';
+
+@Injectable()
+export class UsersService {
+  constructor(private readonly prisma: PrismaService) {}
+
+  async create({ password, ...rest }: CreateUserDto) {
```
Comment (M-Ziusko) [inline backend/users/src/users/users.service.ts:11]:
```suggestion
  async create({ password, ...userWithoutPassword }: CreateUserDto) {
```

File: backend/users/src/users/users.service.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    const SALT_ROUNDS = 10;
+    const hash = await bcrypt.hash(password, SALT_ROUNDS);
+    return this.prisma.user.create({ data: { ...rest, password: hash } });
```
Comment (M-Ziusko) [inline backend/users/src/users/users.service.ts:14]:
```suggestion
    const hash = await bcrypt.hash(password, SALT_ROUNDS);
    
    return this.prisma.user.create({ data: { ...rest, password: hash } });
```

File: backend/users/src/users/users.service.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+import { UpdateUserDto } from './dto/update-user.dto';
+import { PrismaService } from 'src/prisma/prisma.service';
+import * as bcrypt from 'bcrypt';
+
+@Injectable()
+export class UsersService {
+  constructor(private readonly prisma: PrismaService) {}
+
+  async create({ password, ...rest }: CreateUserDto) {
+    const SALT_ROUNDS = 10;
```
Comment (M-Ziusko) [inline backend/users/src/users/users.service.ts:12]: Let's move into constnts folder or env file

File: backend/users/src/users/users.service.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    return this.prisma.user.findUniqueOrThrow({ where: { id } });
+  }
+
+  update(id: string, data: UpdateUserDto) {
+    return this.prisma.user.update({ data, where: { id } });
+  }
+
+  remove(id: string) {
+    return this.prisma.user.delete({ where: { id } });
+  }
```
Comment (M-Ziusko) [inline backend/users/src/users/users.service.ts:31]: Should we return 404 in case non-existing user

File: env.example [hunk 7]
```diff
@@ -0,0 +1,10 @@
+### USERS MICROSERVICE
+USERS_SERVICE_PORT=4000
+
+POSTGRES_USER=users_service
+POSTGRES_PASSWORD=KfTyF7eee
+POSTGRES_PORT=5432
+POSTGRES_DB=users_service_db
+
+USERS_PRISMA_DB_URL="postgresql://users_service:KfTyF7eee@postgres:5432/users_service_db?schema=public"
```
Comment (M-Ziusko) [inline env.example:9]:
```suggestion
USERS_PRISMA_DB_URL="postgresql://users_service:{password}@{user}:{port}/users_service_db?schema=public"
```

File: env.example [hunk 8]
```diff
@@ -0,0 +1,10 @@
+### USERS MICROSERVICE
+USERS_SERVICE_PORT=4000
+
+POSTGRES_USER=users_service
+POSTGRES_PASSWORD=KfTyF7eee
+POSTGRES_PORT=5432
+POSTGRES_DB=users_service_db
```
Comment (M-Ziusko) [inline env.example:7]:
```suggestion
POSTGRES_USER=
POSTGRES_PASSWORD=
POSTGRES_PORT=
POSTGRES_DB=
```

File: package.json [hunk 9]
```diff
@@ -1,10 +1,11 @@
 {
+  "scripts": {
+    "prepare": "husky",
+    "install:all": "npm i && cd backend && npm run install:all"
```
Comment (M-Ziusko) [inline package.json:4]:
```suggestion
    "install:all": "npm i &&  npm  run install:all --prefix backend" 
```

Comment (M-Ziusko) [reply package.json:4]: Or we can use workspace

File: backend/users/src/users/users.service.ts (hunk 6 above)
Comment (foobazdev) [reply backend/users/src/users/users.service.ts:31]: API responses end error types better implement after discussion in future PR and  for all endpoints

File: backend/users/src/prisma/prisma.service.ts (hunk 2 above)
Comment (foobazdev) [reply backend/users/src/prisma/prisma.service.ts]: Done

File: backend/users/src/users/users.service.ts (hunk 3 above)
Comment (foobazdev) [reply backend/users/src/users/users.service.ts]: Done

File: backend/users/src/users/users.service.ts (hunk 4 above)
Comment (foobazdev) [reply backend/users/src/users/users.service.ts]: Done

File: backend/users/src/users/users.service.ts (hunk 5 above)
Comment (foobazdev) [reply backend/users/src/users/users.service.ts]: Done

File: env.example (hunk 7 above)
Comment (foobazdev) [reply env.example]: Done

File: env.example (hunk 8 above)
Comment (foobazdev) [reply env.example]: Done

File: backend/users/src/users/users.service.ts (hunk 6 above)
Comment (foobazdev) [reply backend/users/src/users/users.service.ts:38]: Add exception filter for prisma client

**Review by M-Ziusko — APPROVED** [review]

**Pull request merged**